app/modules/generateWallet.js

```diff
@@ -184,94 +184,133 @@ export const recoverWallet = (wallet: Object): Promise<*> =>
+  return a.WIF
+}
+
+export function validateInputs(
+  wif: string,
+  passphrase: string,
+  passphrase2: string,
+) {
+  if (passphrase !== passphrase2) {
+    throw new Error('Passphrases do not match')
+  }
+  if (!validatePassphraseLength(passphrase)) {
+    throw new Error('Please choose a longer passphrase')
+  }
+  if (wif && !wallet.isWIF(wif)) {
+    throw new Error('The private key is not valid')
+  }
+}
+
+type KeyOption = 'WIF' | 'ENCRYPTED_WIF'
+
 export const generateNewWalletAccount = (
   passphrase: string,
   passphrase2: string,
-  wif?: string,
+  key: string,
+  keyOption: KeyOption,
   history: Object,
   walletName: string,
   authenticated: boolean = false,
   onFailure: () => any = () => undefined,
 ) => (dispatch: DispatchType) => {
-  const isImport = !!wif
   const dispatchError = (message: string) => {
     dispatch(showErrorNotification({ message }))
     return false
   }
-  if (passphrase !== passphrase2) {
-    onFailure()
-    return dispatchError('Passphrases do not match')
-  }
-  if (!validatePassphraseLength(passphrase)) {
-    onFailure()
-    return dispatchError('Please choose a longer passphrase')
-  }
-  if (wif && !wallet.isWIF(wif)) {
-    onFailure()
-    return dispatchError('The private key is not valid')
-  }
+
   const infoNotificationId: any = dispatch(
     showInfoNotification({
       message: 'Generating encoded key...',
       autoDismiss: 0,
     }),
   )
-  return setTimeout(async () => {
-    try {
-      const account = new wallet.Account(wif || wallet.generatePrivateKey())
-      const { WIF, address } = account
-      const encryptedWIF = wallet.encrypt(WIF, passphrase)
-
-      const storedWallet = await getWallet()
-      if (walletName && walletHasLabel(storedWallet, walletName)) {
-        onFailure()
-        return dispatchError('A wallet with this name already exists locally')
-      }
 
-      if (walletHasKey(storedWallet, encryptedWIF)) {
+  let wif = ''
+  // If the key is not given, it means that the user has choosen
+  // the option 'Create Wallet'. Therefore isImport = false.
+  // If the key is given, isImport = true.
+  const isImport = key !== null
+
+  const setWIF = async () => {
+    wif = keyOption === 'WIF' ? key : await decryptEncryptedWIF(key, passphrase)
+    validateInputs(wif, passphrase, passphrase2)
+  }
+
+  setWIF()
+    .then(async () => {
+      try {
+        const account = new wallet.Account(wif || wallet.generatePrivateKey())
+        const { WIF, address } = account
+        const encryptedWIF =
+          keyOption === 'WIF' ? wallet.encrypt(WIF, passphrase) : key
+        const storedWallet = await getWallet()
+        if (walletName && walletHasLabel(storedWallet, walletName)) {
+          onFailure()
+          return dispatchError('A wallet with this name already exists locally')
+        }
+
+        if (walletHasKey(storedWallet, encryptedWIF)) {
+          onFailure()
+          return dispatchError('A  already exists locally')
```

Comment:
What is this error supposed to represent? I believe its supposed to say something like "A wallet with this encrypted key already exists locally."